proposals/plugins.md

```diff
@@ -0,0 +1,146 @@
+being able to merge in the simple changes they need. Examples 
+[here](https://github.com/kubernetes-incubator/kube-aws/issues/340) and 
+[here](https://github.com/kubernetes-incubator/kube-aws/issues/338)
+
+# Specific Examples
+
+* Deployment outside of Kubernetes - Dex uses some additional API server parameter, Datadog agent still has more functionality if not using the containerised version
+* Deployment to Kubernetes - rescheduler, kube2iam
+* Add new parts to CloudFormation - bastion host
+* Customisations to existing parts of CloudFormation - IAM roles/groups
+
+# Plugin Principles
+
+1. Plugin functionality should first be added in places where kube-aws needs it most, i.e. where
+kube-aws receives more feature requests which do not fit with the core feature set.
+1. kube-aws should dogfood the plugins system, i.e. use the plugins system where possible on existing functionality
+1. Everything that is pluggable should be kept as modular as possible, from lists of IAM roles to kube API server parameters.
+1. Where possible, checks will be added to kube-aws to ensure the integrity of the cluster via validation or disallowing particular changes. e.g. disallow overriding node pool init metadata
+   * Note, this could be gradually added rather than forming part of the first implementation
+   * Some of this will likely be provided by CloudFormation but kube-aws could possibly replicas the checks to save time for the stack create/update to fail
+
+# Design
+
+## Plugin Types
+
+* Built-in plugins versus ad-hoc user specific plugins
+  * Some plugins will be provided by kube-aws and some will be specific to a user
+  * Both will use the same mechanisms to hook into kube-aws/deploy etc
+  * Both types of plugins will be enabled and configured in the same way
+* Built-in plugins versus contrib plugins
+  * kube-aws will accept contrib plugins for features considered to be applicable to many users
+  * Contrib plugins will be enabled in a similar way but specifically marked as not fully
+  battle tested; they will still be configured in the same way
+* Plugins will be allowed to act across the whole cluster or act on a specific set of node pools, this should form part of the base configuration for which plugins to run
+
+## Stages
+
+The workflow for kube-aws should consist of the following stages, each of which will be called on the plugins in turn. Each stage will complete across all plugins prior to proceeding to the next stage. All plugins will receive input based on the whole setup, not just their own. This will allow them to hook into parameters from other plugins or core logic. The stages are not intended to become kube-aws commands and should mostly be transparent to a kube-aws user.
+
+1. initialise
+   * Perform any prerequisite setup required such as calls to external system. Configuration, presence of files etc can be validated.
+   * Input - configuration
```

Comment:
how plugins communicate with core? will it be YAML? JSON? binary protocols? which communication channel is used? pipes? socket? passed file descriptors?